cartographer_rviz/cartographer_rviz/submaps_display.cc

```diff
@@ -104,23 +112,28 @@ void SubmapsDisplay::processMessage(
       if (submap_index >= trajectory.size()) {
         trajectory.push_back(
             ::cartographer::common::make_unique<DrawableSubmap>(
-                trajectory_id, submap_index, context_->getSceneManager()));
+                trajectory_id, submap_index, context_->getSceneManager(),
+                submaps_category_, visibility_all_enabled_->getBool()));
+        ConnectVisibilityChange(trajectory.back());
       }
       trajectory[submap_index]->Update(msg->header,
                                        submap_entries[submap_index],
                                        context_->getFrameManager());
     }
+    if (submap_entries.size() < trajectory.size()) {
+      trajectory.resize(submap_entries.size());
+    }
   }
 }
 
 void SubmapsDisplay::update(const float wall_dt, const float ros_dt) {
+  ::cartographer::common::MutexLocker locker(&mutex_);
```

Comment:
Each DrawableSubmap has its own mutex. Line 130 locks the SubmapsDisplay mutex, while QueryOrRenderInProgress() locks the individual DrawableSubmap mutex. Is this a problem?
I have not observed problems with this, and as far as I have tested, you can't crash anymore it by mashing the submaps plugin enable/disable checkbox.